text/0089-tenant-token.md

```diff
@@ -0,0 +1,125 @@
+
+### III. Glossary
+
+| Term               | Definition |
+|--------------------|------------|
+| Master Key         | This is the master key that allows the creation of other API keys. The master key is defined by the user when launching MeiliSearch. |
+| API Key            | API keys are stored and managed from the endpoint `/keys` by the master key holder. These are the keys used by the technical teams to interact with MeiliSearch at the level of the frontend/backend code. |
+| Tenant Token     | These tokens are not stored and managed by a MeiliSearch instance. They are generated for each end-user by the backend code from a MeiliSearch API Key. They are used by the end-users to only search the documents belonging to them. |
+| Multi-Tenancy      | By multi-tenancy, we mean that an end-user only accesses data belonging to him within an index shared with other end-users. |
+
+### IV. Personas
+
+| Persona | Role |
+|---------|------|
+| Mark    | Mark is a developer for a SaaS company. He will implement the code to communicate with MeiliSearch to solve technical/product needs. |
+| UserX   | UserX represents any end-user searching from the frontend interfaces provided by Anna and Mark's company. |
+
+### V. `Tenant Token` Explanations
+
+#### Summary Key Points
+
+- `Tenant tokens` are generated from a MeiliSearch parent `API key` on the user's backend code. They are meant to resolve multi-tenancy by restricting access to data within an index according to the criteria chosen by the team managing the MeiliSearch instance.
+- `Tenant tokens` cannot be less restrictive than their respective parent `API key` and can only be used for searching with a predefined forced filter field.
+- `Tenant tokens` are not stored and thus not retrievable on MeiliSearch. This is why we highly advise setting an expiration time on them for security reasons.
+
+#### Solving Multi-Tenancy with `Tenant tokens`
+
+![](https://i.imgur.com/J4jVe1n.png)
+
+Let's say that `Mark` is a developer for a SaaS platform. He would like to ensure that end-user can only access their documents at search time. **His database contains many users and he hopes to have many more in the future.**
+
+When a user registers, the backend code generates a `Tenant token` specifically for that end-user so they can only access their documents.
+
+The `filter` parameter is set to restrict the search for documents having an `user_id` attribute. This `filter` parameter is contained in the `Tenant token` payload and cannot be modified during the search by the end-user making the request. `filter` can be made of any valid filters. e.g. `user_id = 10 and category = Romantic`
```

Comment:
I understand the scenario for this spec is to restrict searches so an end-user can only search on documents they have access to, but are there other use-cases for tenant tokens?

Reply:
Yes, another possible use-case in the future could be to restrict tokens in terms of rate-limiting for example. That said, the tenants tokens for the moment are generated only for search, it's something we could extend to other actions in the future but in my opinion in the current state of discovery, there is little chance.